### PR TITLE
OM-688 Normal users support

### PR DESCRIPTION
**Pull request by batmahdi**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -93,7 +107,12 @@ export class Index extends Base {
 
     async loadFeatures() {
         try {
-            let [statusData, featuresData] = await Promise.all([this.api.getStatus(), this.api.getFeatures()]);
+            if (this.shared.target === 'cloud') {
+                var statusData = await Promise.all([this.api.getStatus()]);
```
Comment (khenderick) [inline src/index.js]: No need to use `Promise.all` here

File: src/index.js [hunk 2]
```diff
@@ -109,6 +128,17 @@ export class Index extends Base {
         this.signaler.signal('navigate');
     }
 
+    async configAccessChecker() {
+        for (let route of this.router.navigation) {
+            if (route.settings.needGlobalAcl !== undefined && this.shared.installation !== undefined) {
+                route.config.show = this.shared.installation.configurationAccess;
+            } else if (route.settings.group !== 'profile' && this.shared.installation === undefined) {
```
Comment (khenderick) [inline src/index.js]: Does that work on the local gateway interface?

File: src/index.js [hunk 3]
```diff
@@ -259,6 +300,14 @@ export class Index extends Base {
         }, {});
 
         let defaultLanding = this.shared.target === 'cloud' && this.shared.installation === undefined ? 'landing' : Storage.getItem('last');
+        if (!['dashboard', 'outputs', 'thermostats'].contains(defaultLanding) && this.shared.installation !== undefined) {
```
Comment (khenderick) [inline src/index.js]: This doesn't scale well. We need to figure out a way to make this more automatic. However, if the show attribute is correct at this point, I would expect that the below filter checks reset the landing page as expected. Can you verify this?

File: src/index.js [hunk 4]
```diff
@@ -280,6 +329,12 @@ export class Index extends Base {
                             return next.cancel(this.authentication.logout());
                         }
                     }
+                    if (navigationInstruction.config.settings.needGlobalAcl === 'CONFIGURE') {
```
Comment (khenderick) [inline src/index.js]: I've seen this code a few times, is it possible to make this more generic?

For example, the `needGlobalAcl` could be named `needInstallationAccess` which is a list of ACL names. This list, if configured, could then be given to some method on installation's ACL object (see other comments) which could have a `hasAccess(accessKeys)` method which checks whether all keys are allowed.

File: src/index.js [hunk 5]
```diff
@@ -280,6 +329,12 @@ export class Index extends Base {
...
+                        let hasAccess = this.shared.installation === undefined ? false : this.shared.target === 'cloud' ? this.shared.installation.configurationAccess : true
```
Comment (khenderick) [inline src/index.js]: Don't nest ternary operators. You could use something like:

```
let hasAccess  = true;
if (this.shared.target === 'cloud') {
    hasAccess = this.shared.installation === undefined ? false : this.shared.installation.hasAccess(navigationInstruction.config.settings.needInstallationAccess);
}

...
```

File: src/pages/cloud/installations.html [hunk 6]
```diff
@@ -69,7 +69,7 @@ <h3 class="box-title">
                                 </td>
                                 <td click.delegate="checkedChange(installation)">
                                     <span if.bind="installation === shared.installation">${'generic.connected' & t},</span>
-                                    <span if.bind="installation.alive">${'pages.cloud.installations.' + (installation.alive ? 'online' : 'offline') & t}</span>
+                                    <span if.bind="installation.alive !== undefined">${'pages.cloud.installations.' + (installation.alive ? 'online' : 'offline') & t}</span>
```
Comment (khenderick) [inline src/pages/cloud/installations.html]: Can this be `undefined`?

File: src/pages/cloud/installations.js [hunk 7]
```diff
@@ -90,6 +90,18 @@ export class Installations extends Base {
         }
     }
 
+    @computedFrom('allSelectedMain')
+    get hasAtLeastOneConfigAccess() {
+        let hasAccess = false;
+        for (let installation of this.mainInstallations) {
+            hasAccess = hasAccess || installation.configurationAccess;
```
Comment (khenderick) [inline src/pages/cloud/installations.js]: You can do something like

```
if (installation.configurationAccess) {
    return true;
}
```

File: src/pages/dashboard.html [hunk 8]
```diff
@@ -34,12 +34,25 @@ <h4>${'pages.dashboard.nomodules' & t}</h4>
                         <h3 class="box-title">${'pages.dashboard.globalthermostat' & t}</h3>
                     </div>
                     <div class="box-body" style="display: block; margin-right: 10px;">
-                        <span if.bind="globalThermostat === undefined"><i translate.bind="'icons.loading' & t" class="fa fa-spin"></i></span>
-                        <div class="row" if.bind="globalThermostat !== undefined">
+                        <span if.bind="thermostatLoading"><i translate.bind="'icons.loading' & t" class="fa fa-spin"></i></span>
+                        <span if.bind="!thermostatsLoading && !globalThermostatDefined & signal: 'reload-thermostats'">
```
Comment (khenderick) [inline src/pages/dashboard.html]: What do users see that have no access to any thermostat? A normal user should never see a message about where to configure thermostats as he can't configure them anyway.

Similar remark for the thermostat page itself.

File: src/pages/dashboard.html [hunk 9]
```diff
@@ -34,12 +34,25 @@ <h4>${'pages.dashboard.nomodules' & t}</h4>
...
+                                &nbsp;<span class="fa" translate.bind="'icons.right' & t"></span>&nbsp;
+                                ${'pages.settings.thermostats.title' & t}
+                            </a>
+                        </span>
+                        <div class="row" if.bind="shared.target !== 'cloud' && globalThermostatDefined">
                             <div class="col-sm-12">
                                 <global-thermostat thermostat.bind="globalThermostat"></global-thermostat>
                             </div>
                         </div>
+                        <div class="row" if.bind="shared.target === 'cloud' && globalThermostatDefined && !thermostatLoading">
```
Comment (khenderick) [inline src/pages/dashboard.html]: Is this as expected? Here the variable `thermostatLoading` is checked, while it's not on the local gateway interface.

File: src/pages/dashboard.html [hunk 10]
```diff
@@ -70,7 +83,7 @@ <h3 class="box-title">${'pages.dashboard.globalthermostat' & t}</h3>
                     </div>
                 </div>
             </div>
-            <div class="col-lg-3 col-xs-6">
+            <div if.bind="shared.installation.configurationAccess" class="col-lg-3 col-xs-6">
```
Comment (khenderick) [inline src/pages/dashboard.html]: Does this work on the local gateway interface?

File: src/pages/cloud/thermostats.js [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+                }
+                if (!thermostat.isRelay && !this.globalThermostat.isHeating && thermostat.hasCooling) {
+                    thermostats.push(thermostat);
+                }
+            }
+        }
+        return thermostats;
+    }
+
+    getGlobalPreset() {
```
Comment (khenderick) [inline src/pages/cloud/thermostats.js]: Shouldn't this be a getter?

File: src/pages/cloud/thermostats.js [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                if (thermostat.isRelay) {
+                    thermostats.push(thermostat);
+                }
+            }
+        }
+        return thermostats;
+    }
+
+    async loadThermostatUnits() {
+        try{
```
Comment (khenderick) [inline src/pages/cloud/thermostats.js]: Missing space

File: src/pages/cloud/thermostats.js [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+            var data = await this.api.getThermostatUnits();
+            Toolbox.crossfiller(data.data, this.allThermostats, 'id', (id) => {
+                return this.thermostatFactory(id);
+            });
+            for (let thermostat of this.allThermostats) {
+                if (this.globalThermostat.isHeating) {
+                    if (thermostat.hasHeating) {
+                        thermostat.sensorId = thermostat.configuration.heating.sensor_id;
```
Comment (khenderick) [inline src/pages/cloud/thermostats.js]: This needs to be in the container, a page should not have knowledge about such things.

File: src/pages/settings/apps.js [hunk 14]
```diff
@@ -28,6 +28,10 @@ export class Apps extends Base {
         super(...rest);
         this.appFactory = appFactory;
         this.refresher = new Refresher(async () => {
+            if (!this.shared.installation.configurationAccess) {
```
Comment (khenderick) [inline src/pages/settings/apps.js]: This should be covered by a router guard, not by the individual pages.

File: src/pages/cloud/installations.html (hunk 6 above)
Comment (batmahdi) [reply src/pages/cloud/installations.html]: Before running the `checkAlive` call it'll be `undefined`

File: src/containers/cloud/thermostat.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        let id = rest.pop();  // Inverted order
+        super(...rest);
+        this.id = id;
+        this.key = 'id';
+        this.actualTemperature = undefined;
+        this.currentSetpoint = undefined;
+        this.preset = undefined;
+        this.name = undefined;
+        this.roomId = undefined;
+        this._freeze = false;
```
[the remark this answers is not in the document]
Comment (batmahdi) [reply src/containers/cloud/thermostat.js]: Correct

File: src/pages/settings/apps.js (hunk 14 above)
Comment (batmahdi) [reply src/pages/settings/apps.js]: There are two scenarios:
- navigating between pages: handled by the router guard
- changing between installations (possible different permissions): here handled by the refresher method that runs when the installation is updated.

But on second thought, the code will be cleaner to implement this in `base.js` at `installationUpdated` method

File: src/pages/cloud/thermostats.js (hunk 13 above)
Comment (batmahdi) [reply src/pages/cloud/thermostats.js]: In the container, a thermostat has no visibility over the thermostat-group's current mode. That's why it's done here.

File: src/index.js (hunk 3 above)
Comment (batmahdi) [reply src/index.js]: Without this if statement, if the user's role has changed and the last page he visited was a settings page, the next time the user opens the portal he will get redirected to the `cloud/nopermission` page.

File: src/pages/dashboard.html (hunk 9 above)
Comment (batmahdi) [reply src/pages/dashboard.html]: But it is, see `dashboard.js:135` and `dashboard.js:145`.

File: src/locales/en/translation.json [hunk 16]
```diff
@@ -858,6 +858,9 @@
                 "title": "Inputs",
                 "usedbygroupaction": "Controlled by GroupAction"
             },
+            "nopermission": {
+                "message": "You have no permission to view this page. Contact your administrator."
```
Comment (khenderick) [inline src/locales/en/translation.json]: Let's use a slightly more descriptive message and use a title and message to make it look a tad less "in your face".
* Title: "You have no access"
* Message: "You have no access to this page. If you think this is by mistake, contact the person that created your account."

It also makes it a bit more clear who they have to contact; this will typically be a family member in a residential scenario, or e.g. the facility manager in an office building.

Dutch:
* Title: "Geen toegang"
* Message: "Je hebt geen toegang tot deze pagina. Als je denk dat dit een vergissing is, neem dan contact op met de persoon die jouw account heeft gemaakt."

File: src/pages/cloud/installations.js [hunk 17]
```diff
@@ -90,6 +90,16 @@ export class Installations extends Base {
         }
     }
 
+    @computedFrom('allSelectedMain')
```
Comment (khenderick) [inline src/pages/cloud/installations.js]: Is this the correct field for `computedFrom`? It feels unrelated.

File: src/pages/dashboard.html (hunk 9 above)
Comment (khenderick) [reply src/pages/dashboard.html]: I was only referring to `dashboard.html:49` where it doesn't take `thermostatLoading` into account. This is because it makes sense to only show this control when `thermostatLoading` is false, as otherwise you might briefly see the loading indicator _and_ the control.

File: src/pages/cloud/thermostats.js (hunk 13 above)
Comment (khenderick) [reply src/pages/cloud/thermostats.js]: Yep, you're right. Damn, the API is really lacking some important information.

File: src/resources/base.js [hunk 18]
```diff
@@ -60,5 +60,11 @@ export class Base {
     }
 
     installationUpdated() {
+        if (this.shared.installation !== undefined) {
```
Comment (khenderick) [inline src/resources/base.js]: Two questions here:
* Is this code still executed if it is implemented in the pages themself and if so, what's the use of this?
* Is this desired? It looks like if the customer is on the output page, and he switches to an installation where he has no configuration access to, he'll be redirected to the dashboard, even though he still has access to the output page.

File: src/pages/settings/apps.js (hunk 14 above)
Comment (khenderick) [reply src/pages/settings/apps.js]: I see what you mean, this indeed makes sense. However, a few questions on the implementation of `installationUpdated`.

File: src/index.js (hunk 3 above)
Comment (khenderick) [reply src/index.js]: Even though this code seems to be updated; the thing is that inside `configAccessChecker` the route's `show` config entry is set to false in case a user isn't allowed to see it. A few lines below this change there's a `if (routes.filter(route) => route.show === true ...` line that - if I recall correctly - will make sure that the `defaultLanding` is set to a page that can be shown. So this code should also kick in if the user has no access, right?

File: src/pages/dashboard.html (hunk 9 above)
Comment (batmahdi) [reply src/pages/dashboard.html]: `thermostatLoading` will be true at init time, once the the thermostats are loaded it'll be false. It'll avoid showing the loading icon every now and then.

File: src/resources/base.js (hunk 18 above)
Comment (batmahdi) [reply src/resources/base.js]: You're right I haven't noticed the side effect of moving the code to `base.js`.

But something similar must be implemented to handle the case of changing to installation that results in a change of permissions, admin -> normal and the last visited page is a configuration page.

File: src/index.js [hunk 19]
```diff
@@ -128,13 +128,20 @@ export class Index extends Base {
...
         if (this.shared.target === 'cloud') {
-            for (let route of this.router.navigation) {
+            for (let route of routes) {
+                if (route.settings === undefined) {
+                    continue;
+                }
                 if (route.settings.needInstallationAccess !== undefined && this.shared.installation !== undefined) {
-                    route.config.show = this.shared.installation.configurationAccess;
+                    if (route.show !== undefined) {
+                        route.show = this.shared.installation.configurationAccess;
```
[the remark this answers is not in the document]
Comment (batmahdi) [reply src/index.js]: If the provided routes is just the raw array of objects, it'll only be `route.show`. If the routes are Aurelia's configured routes and no longer just a raw array of objects, it'll have a config attribute: `route.config.show`.
